File: 1_Reverse_Words_In_A_Sentence/reversewords.cpp
#include<iostream>
using namespace std;
#include<string.h>

void reverseString(char* str, int length)
{
	if (str == NULL)
	{
		return;
	}
	if (length<=1)
	{
		return;
	}
	char *p=NULL;
	p = new char[length];
	strncpy(p, str, length);
	int i;
	for (i=0; i<length;i++)
	{
	str[i]=p[length-i-1];
	}
	delete p;
	return;
}

void reverseWithSpaces(char* str, int length)
{

	if( str == NULL)
		return;
	if (length<=1)
	{
		return;
	}
	int i, start=0;
	for(i=0;i<length;i++)
	{
	if(str[i]==' ')
	{
		start++;
		continue;
	}
	if (str[i+1]==' '||str[i+1]=='\0')
	{
		if(i>0)
		{
		reverseString(str+start,i-start+1);
		start=i+1; 
		}
		else
		{
		start++;
		}
	}
	
	}
	reverseString(str+start,length-start+1);
}

void reverseWords(char* str) //, int length) Usually pass the second argument as length
{
	if (str == NULL)
	{
		cout <<"String is NULL"<<endl;
		return; 
	}
	int length = strlen(str);
	cout<<"String to be reversed is : "<<str<<endl;
	cout<<"Length of the string is:" <<length<<endl;
    //1. Reverse the whole string
	reverseString(str, length);
    cout<<"Reverse string is : "<<str<<endl;
    //2.With the reverse string reverse each string within the spaces
    reverseWithSpaces(str, length);
    cout<<"Reverse sentence is : "<<str<<endl;
    
}

int main()
{
	char *p = NULL;
	//reverseWords("Hello World");
	///reverseWords("");
	reverseWords(p);
	p = new char[100];
	strcpy(p,"Hello");
	reverseWords(p);
	strcpy(p,"Hello World");
    reverseWords(p);
    strcpy(p,"This is   done");
    reverseWords(p);
	delete p;
}

File: 1_Reverse_Words_In_A_Sentence/ReverseWords.java
import java.util.*;

class ReverseWords
{
    
    public static char[] reverseString(char str[])
	{
		if (str == null)
		{
			System.out.println("String is null in Function");
			return null;
		}
		if (str.length<=1)
		{
			return str;
		}
		
		int length = str.length;
		char rev[] = new char[length];// This is important else there will be problems if memory is not allocated properly
		int i=0;
		//System.out.println(str.length);System.out.println(str);
		for (i=0;i<str.length; i++)
		{
		rev[i]= str[length-i-1];
		//System.out.println(str[length-i-1]);
		//System.out.println(length -i-1);
		}
		
		return rev;
    }
    
	public static char[] reverseSentence(char str[])
	{
		if (str == null)
		{
			System.out.println("String is null in Function");
			return null;
		}
		if (str.length<=1)
		{
			return str;
		}
		
		str = reverseString(str);
	    System.out.println("Reversed String is "+ new String(str));
	    int i;
	    //String[] stringList = new String[str.length];
	    int j;
	    int start=0;
	    String finalString= "";
	    for(i=0;i<str.length;i++)
	    {
	        if(str [i]==' ')
	        {
	        start++;
	        finalString = finalString + " ";
	        continue;
	        }
	            if(i+1==str.length||str[i+1]==' ')
	            {
	            char newarray[] = new char [i-start+1];
	            int k=0;
	            for (j=start;j<=i;j++,k++)
	            {
	            newarray[k]=str[j];
	            }
	            newarray= reverseString(newarray);
	            finalString = finalString + new String(newarray);
	            System.out.println("Reversed String is "+ new String(newarray));
                start =i+1;
	            } 
	        

	    }

		/*
		char[] p = new char[str.length];
		strncpy(p, str, length);
		int i;
		for (i=0; i<length;i++)
		{
		str[i]=p[length-i-1];
		}
		delete p;
		return;
		*/
		return finalString.toCharArray();
	}
	
	public static void main(String args[])
	{
		char str[]= null;
		String s = "Hello  World";
		str = new char[s.length()];
		str = reverseString(str);
		if(str== null)
		{
		System.out.println("String is null");
		}
		str = s.toCharArray();
		System.out.println(str.length);
		str = reverseString(str);
		System.out.println("Reversed String is "+ new String(str));
		str = new char[s.length()];
		str = s.toCharArray();
		str = reverseSentence(str);
		System.out.println("Reversed Sentence is "+ new String(str));
	}
}



/*
Solutions

https://stackoverflow.com/questions/2713655/reverse-a-given-sentence-in-java


Approach 1:
String[] words =  "This is interview question".split(" ");\

String rev = "";
for(int i = words.length - 1; i >= 0 ; i--)
{
   rev += words[i] + " ";
}

// rev = "question interview is This "


Approach 1.2:
// can also use StringBuilder:
StringBuilder revb = new StringBuilder();
for(int i = words.length - 1; i >= 0 ; i--)
{
   revb.append(words[i]);
   revb.append(" ");
}
// revb.toString() = "question interview is This "

Or Use:

String[] words = sentence.split(" ");
String[] reversedWords = ArrayUtils.reverse(words);
String reversedSentence = StringUtils.join(reversedWords, " ");

/*


